contracts/brush/proc_macros/lib.rs

```diff
@@ -264,45 +254,140 @@ pub fn storage_trait(_attrs: TokenStream, _input: TokenStream) -> TokenStream {
+/// where this macro will be used. Also you must understand, that `return` can break other modifiers
+/// (the method can have several modifiers).
+///
+///  ** Note ** This macro must be processed before [`#[brush::modifiers]`](`macro@crate::modifiers`),
+/// otherwise it will fail: It means that [`#[brush::modifier_definition]`] must be defined in scope of
+/// [`#[brush::contract]`](`macro@crate::contract`)
+/// or it must be defined in another crate(macros in dependencies will be processed early).
+///
+/// # Example: Definition
+///
+/// ```
+/// #[derive(Default)]
+/// struct Contract {
+///     initialized: bool,
+/// }
+///
+/// impl Contract {
+///     #[brush::modifier_definition]
+///     fn once(&mut self) {
+///         assert!(!self.initialized, "Contract already is initialized");
+///         #[body]();
+///         self.initialized = true;
+///     }
+/// }
+/// ```
 #[proc_macro_attribute]
-pub fn modifiers(_attrs: TokenStream, method: TokenStream) -> TokenStream {
-    let attrs: TokenStream2 = _attrs.into();
-    let modifiers = attrs
-        .into_iter()
-        .filter_map(|token|
-            if let TokenTree::Ident(ident) = token {
-                Some(ident)
-            } else {
-                None
-            })
-        .collect();
-
-    let fn_item = syn::parse2::<ItemFn>(method.clone().into());
-    let trait_method_item = syn::parse2::<TraitItemMethod>(method.clone().into());
-
-    let mut code: TokenStream2 = method.into();
-    if let Ok(mut item) = fn_item {
-        add_modifiers_to_block(&mut item.block, modifiers);
-        code = quote! { #item };
-    } else if let Ok(mut item) = trait_method_item {
-        if let Some(block) = &mut item.default {
-            add_modifiers_to_block(block, modifiers);
-            code = quote! { #item };
-        }
-    }
-
-    code.into()
+pub fn modifier_definition(_attrs: TokenStream, _input: TokenStream) -> TokenStream {
+    modifier_definition::generate(_attrs, _input)
 }
 
-#[inline]
-fn add_modifiers_to_block(block: &mut syn::Block, modifiers: Vec<syn::Ident>) {
-    modifiers
-        .into_iter()
-        .for_each(|ident| {
-            let code = quote! {
-                #[cfg(not(feature = "ink-as-dependency"))] self.#ident();
-            };
-            block.stmts.insert(0, syn::parse2::<syn::Stmt>(code)
-                .expect("Can't parse statement of modifier"));
-        });
-}
+/// Macro pastes the code from the modifier definition inside of the function.
+/// It means that every stuff from the modifier definition must be available in the scope of marked method.
+///
+/// Modifiers are designed to be used for methods marked with the `#[ink(message)]` attribute.
+/// You can try to use them in internal implementations and foreign functions, but you must be sure,
+/// that [`#[brush::modifier_definition]`](`macro@crate::modifier_definition`) will be processed early.
```

Comment:
```suggestion
/// You can try to use them in internal implementations and foreign functions, but you must be sure,
/// that [`#[brush::modifier_definition]`](`macro@crate::modifier_definition`) will be processed earlier.
```

Reply:
what are foreign functions? 

I guess it's better to use some other word, because "foreign functions" sounds a lot like FFIs, which doesn't make sense in this context